### DISCUSSION TITLE
How I can merge two different coverage folder?

### DISCUSSION THREAD

**Discussion by AgarkovRoman** [vitest-dev/vitest#3744] [Q&A] [answered]
I have a few packages and each of them has own coverage folder. I want to merge them 
How I can do it?
Or maybe there is a better solution? (To have one coverage folder and update it with every new package coverage)

Btw, I am using Istanbul

**Comment by AriPerkkio** [comment, marked as answer]
If the packages are in a monorepo you should use [Vitest's workspaces](https://vitest.dev/guide/workspace.html). It will merge the coverage reports out-of-the-box for whole project.

If these are completely separate projects and for some reason you want to merge the coverage reports, you could use `nyc` for it, e.g. `nyc merge`: https://github.com/istanbuljs/nyc#combining-reports-from-multiple-runs

**Reply by fkrauthan** [reply]
@AriPerkkio Quick follow up question. To use `nyc merge` you need the raw coverage file generated by istanbul. However it looks like vitest does not write that file into the standard `.nyc_output` folder. So where do I get the raw coverage report file from?

**Reply by AriPerkkio** [reply]
The `.nyc_output` contains JSON file created by `json` reporter: https://vitest.dev/config/#coverage-reporter.

If you want to use same directory name with Vitest, I think following might work:

```js
coverage: {
  reportsDirectory: ".nyc_output",
  reporter: [ ["json", { file: "coverage.json" }] ],
}
```

**Reply by fkrauthan** [reply]
@AriPerkkio The json report file is a summary so it can't be used to properly combine multiple test runs (e.g. a unit test run via vitest and a integration test run via playwright). What you would need is the raw JSON instanbul generates (the one that gets passed to the report generators).

**Reply by AriPerkkio** [reply]
I think this one is it. 

When you are merging reports from Playwright, make sure you are not mixing Istanbul and V8 reports. Use Istanbul on Playwright (instead of Chromium's V8) when using Istanbul on Vitest. 

Here's equivalent guide for Jest: https://rafaelalmeidatk.com/blog/merging-coverage-reports-from-jest-and-cypress. Looks like it's using same `json` reporter.

**Reply by fkrauthan** [reply]
Oh sorry never mind somehow when I quickly looked at it it looked completely different but I should have looked at the full file. Yeah I think using `json` reporter should work.

**Reply by alanskovrlj** [reply]
@AriPerkkio I'm trying to achieve the same result using [Vitest's workspaces](https://vitest.dev/guide/workspace.html), but it still seems to generate separate coverage reports for each package instead of merging them and I couldn’t find anything related to this in the documentation. Are there any examples where this is achieved out of the box?

**Reply by AriPerkkio** [reply]
@alanskovrlj here's one example https://stackblitz.com/~/edit/vitest-workspace-coverage

**Reply by alanskovrlj** [reply]
@AriPerkkio Thanks! My mistake, I realized I was running vitest command for each package in a loop instead from monorepo root. Anyway, works nicely now :smiley:
